# "Should you use synthetic data for label balancing"

After having worked with tabular synthetic data for the past 6 months, I have encountered many articles claiming that synthetic data is the solution for many (machine learning) problems. While synthetic data can serve as a useful Privacy Enhancing Technology (PET) and has shown to be useful in certain tasks, its usefulness and relevance is not always clearly assessed. An example of this is, and also the inspiration for me writing this article, is the article provided by Synthetic Data Vault (SDV) titled: "Can you use synthetic data for label balancing?" (https://sdv.dev/blog/synthetic-label-balancing/).

The article addresses a well-known issue in classification problems: imbalanced target labels. The article correctly identifies techniques like Random Oversampling (ROS) and noise injection while acknowledging their downsides (being overfitting and noisy data). However, the article then presents synthetic data as a "compelling solution" without providing enough evidence of this. While I am a fan of SDV, their generators, preprocessors, and constraints, this article overlooks important aspects validating the validity of synthetic data for these problems. Although you definitely can use synthetic data for label balancing (to answer the question of the article), the key question is whether you **should** use synthetic data and how it compares to state-of-the-art (SOTA) techniques.

Throughout this article, I aim to provide an answer to this question by comparing synthetic data produced by SDV generators against alternatives and build on top of the aformentioned article. Specifically, I compare data-level approaches such as noise injection, ROS, Synthetic Minority Over-sampling TEchnique (SMOTE), and CTGAN against the algorithm-level approach of Cost-Sensitive learning. This idea is not novel and adjacent research is available in literature. Adiputra and Wanchai (2024) for instance compare similar approaches resampling and synthetic data approaches. However, their calidation approach uses cross validation while resampling before cross validation, which is a common pitfall leading to data leakage. Also a mistake in section 5.7.4 of: https://d2l.ai/chapter_multilayer-perceptrons/kaggle-house-price.html. It is not the exact same mistake, but similar. Transformations should be applied separately so doesnt really

This article aims to improve on this by providing a more methodologically sound approach whilst providing the intuition and explanation for practictioners that are less familiar with imbalanced classification problems.

### Notes

- Check the helper_functions
    - move _get_num_samples to init?
    - Fix scoring function for results
    - Should I use fit/transform as opposed to fit_resample/resample in the custom samplers


## Imports

Notably, to ensure a proper cross validation procedure, I use a pipeline. Specifically, the pipeline from imb_learn is used over sklearn's

In [11]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt     
import seaborn as sns

import lightgbm as lgb

from imblearn.over_sampling import RandomOverSampler, SMOTE
from imblearn.pipeline import Pipeline

from sklearn.dummy import DummyClassifier
from sklearn.metrics import f1_score, confusion_matrix, classification_report, roc_auc_score, precision_score, recall_score
from sklearn.model_selection import StratifiedKFold, cross_validate, train_test_split

from sdv.metadata import Metadata
from sdv.single_table import CTGANSynthesizer, TVAESynthesizer

from helper_functions import NoiseSampler, ColumnScaler, SDVSampler, display_scores

In [2]:
import warnings
warnings.filterwarnings("ignore")

## Data exploration

For this analysis, the creditcard dataset will be used from Kaggle (https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud), containing transactions and whether they were fraudulent or not. The goal is to predict whether a transaction is fraudulent or not, making it a classification task. Naturally, the amount of non-fraudulent transactions outweigh the number of fraudulent transactions resulting in an imbalanced classification task.

For this analysis, we will only be using a subset of the columns, which in this case contains mostly the Principal Components of the original data for confidentiality reasons. In addition, we also have a variable called "Time" which is the seconds from the first transaction in the dataset, 'Amount' which is the amount spent on the transaction, and 'Class' indicating whether the transactions is fraudulent or not, which also is our target variable. Ultimately, the data consists mostly of floats with the exception of our target which is binary.

In [3]:
# Select first then and last 2 columns
creditcard = pd.read_csv('../data/creditcard.csv')

# To reduce the dimensiolality of the dataset, we will only use the first 12 principal components,
# Time, Amount, and target
print(creditcard.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

Next, we see how our target labels are distributed.

In [4]:
# Target label distribution rounded to 2 decimal places
round(creditcard['Class'].value_counts(normalize = True) * 100, 2)

Class
0    99.83
1     0.17
Name: proportion, dtype: float64

It becomes evident that the dataset is highly imbalanced, with  only 0.17% of transactions being fraudulent. This can be problematic for machine learning algorithms if you were to not account for this as models tend to bias towards the non-fraudulent cases given their overrepresentation in the data. As a result, a model predicting every transaction to be non-fraudulent would already result in a 99% accuracy even though every fraudulent transaction has been wrongfully predicted. This highlights the importance of carefully selecting evaluation metrics, as accuracy alone can be misleading in imbalanced classification problems.

### Data splitting

Next step is to split the data. We stratify on the target variable to ensure an even split across train/set sets.

In [10]:
X = creditcard.drop('Class', axis = 1)
y = creditcard['Class']

## Metrics

As mentioned previously, the choice of metrics is far from trivial. However, given that it is beyond the scope of this blog, I will choose an F1-score without diving too deep into the costs of misclassification.

### Algorithm

For the algorithm, I will use a LGBM Classifier. Choosing the most optimal estimator is beyond the scope of this blog. LGBM is chosen for its efficiency and relative predictive power, therefore being used consistently.

## Settings

We define the folds and parameters to optimize over as these will be consistent across resampling techniques.

In [6]:
RANDOM_STATE = 2

CV_FOLDS = StratifiedKFold(n_splits = 5, random_state = RANDOM_STATE, shuffle = True)

scorings = ['f1', 'roc_auc', 'precision', 'recall']

## Incorrect cross validation procedure

Suppose we wish to use ROS to oversample the fraud instances. A common mistake that is made is to perform this oversampling before splitting the data. Why is this wrong: "Animation here". To illustrate this we will test two predictive models on truly unseen data. Therefore, for this section we will make use of a train test split. The train set is used to determine the cross validation scores. Next, the cross validation scores are compared. Next predictions are made on the test set and then the predictions on the test set are compared.



The cross validation mistake is easily made and looks like this implemented in python:

In [13]:
# Split the dataset into training and testing
X_cv, X_test, y_cv, y_test = train_test_split(X, y, test_size = 0.2, stratify = y,
                                                    random_state = RANDOM_STATE)

In [ ]:
# We resample the data used for cross-validation
ros = RandomOverSampler(random_state = RANDOM_STATE)
X_res, y_res = ros.fit_resample(X_cv, y_cv)

cv_score = cross_validate(
    lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1),
    X_res, y_res, cv = CV_FOLDS, scoring = scorings, n_jobs = -1
)

cv_score = cv_score['test_f1'].mean()
print(f"Cross validation score: {cv_score:.4f}")

Cross validation score: 0.9999


This approach leads to (near) perfect scores. Hopefully, this raises some suspicison as to the validity of the results. This seems great, but how do these results translate to truly unseen data: the test set we have not used in the RandomOverSampler.


In [16]:
# We define the model and fit it to the resampled cv data
lgbm_classifier = lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, 
                                     verbose = -1)
lgbm_classifier.fit(X_res, y_res)

# predictions on the unseen test data and evaluation
preds = lgbm_classifier.predict(X_test)
test_score = f1_score(y_test, preds)
print(f"Test score: {test_score:.4f}")


Test score: 0.8442


We can see that the test results are lower than the cross validation results indicating that the cross validation score in this way is not representative. Typically you also want to tune hyperparameters during cross validation using GridSearch for example. In this case, this might lead to even worse results because hyperparameters are chosen during incorrect cross validation procedure and fit to that which might results in worse generalization.

Although apparent in this situation given the near perfect CV scores, mostly as a results of the severe class imbalance (many observations are duplicates), this won't always be the case with every dataset. In addition, there are other ways of data leakage that are more subtle (explain feature engineering ( including male/female --> 1/0)). Best practice to do everything for train/test separately

#### Random OverSampling the Correct Way

Next up, we perform cross validation the correct way. There are multiple ways to do it. Namely, iterating over each fold in a loop and manually applying the required transformations to the train and holdout set within each fold separately or using a pipeline. For simplicity sake, we will be using the latter option.

In [17]:
# Define the sampling pipeline
pipeline_ros = Pipeline(
    [('ros', RandomOverSampler(random_state = RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

# Now we apply the cross-validation on the pipeline using the non resampled data
# Subsequently the pipeline will resample the data for each fold separately
cv_score_ros = cross_validate(pipeline_ros, X_cv, y_cv, cv = CV_FOLDS, scoring = scorings, 
                               n_jobs = -1)

cv_score_pipe = cv_score_ros['test_f1'].mean()
print(f"Cross validation score with pipeline: {cv_score_pipe:.4f}")

Cross validation score with pipeline: 0.8384


Although the score is less impressive than the first method, it more accurately reflects the results on the holdout set of 0.8442. Furthermore, (I forgot what I wanted to say)

## Why does this happen? What exactly is going wrong?

Compare and display duplicates. I can display duplicates through manual cross validation procedure (before and after)

## Baseline performance

As a baseline for this problem to compare predictions to, we will use the DummyClassifier from sklearn. We use 'stratified' becasue choosing 'most_frequent' is probably also a natural choice, but given that we mostly evaluate on F1, all scores would be zero, making the improvement over the baseline seem better more impressive.

#### Perhaps move this section to later? after the explanation of what goes wrong for instance?

In [ ]:
pipeline_baseline = Pipeline(
    [('dummy', DummyClassifier(strategy = 'stratified', random_state=RANDOM_STATE)),]
)

cv_score_base = cross_validate(pipeline_baseline, X, y, cv = CV_FOLDS, scoring = scorings, 
                                n_jobs = -1)


## Random OverSampling

Now on the entire dataset during cross validation instead of a split

In [ ]:
# Define the sampling pipeline
pipeline_ros = Pipeline(
    [('ros', RandomOverSampler(random_state = RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

cv_score_ros = cross_validate(pipeline_ros, X, y, cv = CV_FOLDS, scoring = scorings, 
                               n_jobs = -1)

## Noise injection

The first approach mentioned in the article is noise injection. Although I have not seen it being used in practice and the article does not mention the noise generating process, a uniform sampling procedure will be used. Specifically, for each variable I will extract their minimum and maximum values and use them to sample from a uniform distribution. Therefore, the correlations between variables are overlooked and the bivariate distributions won't be correct.

Given that all variables, with exception of the target variable, are numerical this step is quite straightforward.

In [10]:
# Define the sampling pipeline
pipeline_noise = Pipeline(
    [('noise_sampler', NoiseSampler(random_state = RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

cv_score_noise = cross_validate(pipeline_noise, X, y, cv = CV_FOLDS, scoring = scorings, 
                                n_jobs = -1)

## SMOTE

In [12]:
# Define the sampling pipeline
pipeline_smote = Pipeline(
    [('scaler', ColumnScaler(['Amount', 'Time'])),
     ('smote', SMOTE(random_state = RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

cv_score_smote = cross_validate(pipeline_smote, X, y, cv = CV_FOLDS, scoring = scorings, 
                                n_jobs = -1)


## SDV Generators

The SDV generators that will be compared are the CTGAN and TVAE

In [13]:
metadata = Metadata.detect_from_dataframe(X)

### CTGAN

I will use the CTGAN synthesizer in this case with default parameters and without any constraints.

In [14]:
# Define the sampling pipeline
pipeline_ctgan = Pipeline(
    [('ctgan', SDVSampler(CTGANSynthesizer, metadata, random_state= RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

cv_score_ctgan = cross_validate(pipeline_ctgan, X, y, cv = CV_FOLDS, scoring = scorings, 
                               n_jobs = -1)

## TVAE Synthesizer

I will also use the TVAE Synthesizer. Again, there will be no hyperparameter tuning and no constraints.

In [15]:
# Define the sampling pipeline
pipeline_tvae = Pipeline(
    [('tvae', SDVSampler(TVAESynthesizer, metadata, random_state= RANDOM_STATE)),
     ('LGB', lgb.LGBMClassifier(random_state = RANDOM_STATE, n_jobs = -1, verbose = -1))]
)

cv_score_tvae = cross_validate(pipeline_tvae, X, y, cv=CV_FOLDS, scoring = scorings,
                               n_jobs = -1)

## Cost sensitive learning

I will be using the inverse class frequency for this as the real costs associated with misclassifying creditcard fraud is not known in this example. These weights will be specified using the classifier's 'class_weight' parameter. Specifically, this is set to 'balanced' to achieve the inversely proportional weights. 

Given that the folds are stratified, we assume the that the assigned weights for the entire train set is (roughly) equal to the weights for each fold. This will result in the following weights being assigned:


In [16]:
X.shape[0] / (2 * np.bincount(y)) 

array([  0.50086524, 289.43800813])

The pipeline then becomes

In [17]:
# Define the sampling pipeline
pipeline_csl = Pipeline(
    [('LGB', lgb.LGBMClassifier(class_weight= 'balanced', random_state = RANDOM_STATE,
                                n_jobs = -1, verbose = -1))]
)

cv_score_csl = cross_validate(pipeline_csl, X, y, cv=CV_FOLDS, scoring = scorings,
                               n_jobs = -1)

# Results

In [18]:
print(cv_score_csl['test_f1'].mean())
cv_score_csl

0.8388369280184002


{'fit_time': array([3.92413664, 3.97503018, 4.17920709, 3.91475725, 4.23648381]),
 'score_time': array([0.27906108, 0.24668121, 0.18132877, 0.25999761, 0.16204786]),
 'test_f1': array([0.86458333, 0.83333333, 0.83248731, 0.83838384, 0.82539683]),
 'test_roc_auc': array([0.98446967, 0.97377153, 0.9660105 , 0.9700833 , 0.98033602]),
 'test_precision': array([0.89247312, 0.80952381, 0.82828283, 0.83      , 0.85714286]),
 'test_recall': array([0.83838384, 0.85858586, 0.83673469, 0.84693878, 0.79591837])}

In [19]:
scorings

['f1', 'roc_auc', 'precision', 'recall']

In [25]:
cv_score_ros

{'fit_time': array([8.1566565 , 8.48002768, 8.28060746, 8.15264034, 8.4801054 ]),
 'score_time': array([0.25648475, 0.18453979, 0.19741416, 0.2393806 , 0.17600703]),
 'test_f1': array([0.86458333, 0.84102564, 0.83597884, 0.81818182, 0.82608696]),
 'test_roc_auc': array([0.98166895, 0.97387678, 0.9725407 , 0.96814955, 0.97856439]),
 'test_precision': array([0.89247312, 0.85416667, 0.86813187, 0.81      , 0.88372093]),
 'test_recall': array([0.83838384, 0.82828283, 0.80612245, 0.82653061, 0.7755102 ])}

In [ ]:
cv_scores = [cv_score_base, cv_score_noise, cv_score_ros, cv_score_smote,
             cv_score_ctgan, cv_score_tvae, cv_score_csl]   

In [28]:
# Complete the scores of 

r = display_scores(cv_scores, scorings)
r

f1  roc_auc  precision  recall
Baseline  0.000    0.499      0.000   0.000
Noise     0.744    0.974      0.683   0.823
ROS       0.837    0.975      0.862   0.815
SMOTE     0.654    0.961      0.538   0.837
CTGAN     0.760    0.975      0.718   0.821
TVAE      0.757    0.972      0.697   0.831
CSL       0.839    0.975      0.843   0.835